docs/EVALUATION.md

```diff
@@ -0,0 +1,249 @@
+
+##### More fine grained view
+
+Accounting for the afore mentioned assumptions that training takes some time and triggers are not at one point in time but specify a training batch (time interval), we can visualize a pipeline run as follows:
+
+
+```mermaid
+gantt
+    title Pipeline run
+    dateFormat YYYY-MM-DD
+    axisFormat %Y
+    section models
+        Trigger1 :milestone, t1, 1930, 0d
+        train_batch_1 :a10, 1928, 2y
+        training1 :a11, after t1, 1y
+        model 1          :a12, after a11, until a22
+
+        Trigger2 :milestone, t2, 1940, 0d
+        train_batch_2 :a20, after a10, until t2
+        training2 :a21, after t2, 1y
+        model 2          :a22, after a21, until a32
+
+        Trigger3 :milestone, t3, 1945, 0d
+        train_batch_3 :a30, after a20, until t3
+        training3 :a31, after t3, 1y
+        model 3          :a32, after a31, until a42
+
+        Trigger4 :milestone, t4, 1950, 0d
+        train_batch_4 :a40, after a30, until t4
+        training4 :a41, after t4, 1y
+        model 4          :a42, after a41, 10y
+
+```
+
+With our assumption of `training_time=0` this `training_<i>` would be can be removed so that we can evaluate one model on the data between two triggers (hence the next training batch)
+
+##### State of modyn (at the time of writing)
+
+We currectly evaluate a model directly after training it (e.g. we are now at `Trigger 4`) and have now robust approach of combining the results of the single model evaluations to a pipeline performance. We support a `OffsetEvaluationStrategy` that specifies the interval in which we want to evaluate the models. We also support `MatrixEvaluationStrategy`: We evaluate on predefined grid of evaluation points.
+
+We face several problems:
+
+1) We don't know when the next trigger occurs.
+    This might be fine if we assume a time based trigger policy with a regular schedule where we can set the OffsetEvaluationStrategy lookahead offset to the trigger interval. However, for a generic trigger policy we cannot assume this. In consequence when setting a fixed lookahead offset we might end up evaluating a model with a lookahead offset that is too large or too small. (E.g. when always evaluate on the next 5 years after a training. That would mean that if a policy doesn't retrain for 20 years we miss a lot of data)
+
+> After `Trigger 4` we e.g. could end up only considering the first 5 years of the time where `model 4` is used. That could completely skew the pipeline performance if e.g. as heavy drift occurs after 5 years of using `model 4`.
+
+2) As mentioned above the individual performances of every evaluation cannot simple be averaged to get the pipeline performance as number of samples in every time range is important!
+
+3) Without making any assumptions on how & when to trigger `MatrixEvaluationStrategy` is not of any use as the grid splits are arbitrary and don't correspond to the times when triggers occur and models are transitioned.
```

Comment:
I'd not necessarily agree here. Let's consider CLOC again. The simple case is that we have the `n*m` score matrix where we trigger each year and evaluate on each year. Now, let's consider e.g. a drift policy where we don't know before when we trigger. Let's say we trigger x times. We can still build up a `x*m` matrix, that tells us for each produced model, how well it performs on each year?

And if this is not helpful, imagine a plot like in the Modyn paper. On the x-axis we take the evaluation year and on the y axis, we plot the accuracy of the _current_ model that is most recent for each year on the x-axis. This can be extracted from the matrix easily. If the drift strategy triggers a lot less frequently than yearly triggering, but has a similar accuracy curve, we'd easily see that it can be deemed "better".

Overall, I think having _fixed evaluation time slices_ actually is wanted and useful, because it gives a common reference point for all types of triggering strategies where we can ask: How well is the current most recent model in the pipeline performing (by whatever metric)?

Reply:
I have considered that already and it makes sense, yes! I wasn't able to judge how computationally expensive it would be to evaluate all models on all slices (obviously that depends on the number of slices).
In addition to that I had a slight modification to this idea in mind. Instead of using partitions of the dataset ("slices"), it might make sense to use some kind of rolling time window (with the evaluation time in the window's center). By doing we could estimate the performance at a point `t` by considering `offset` years into the past and future. The difference to slicing is that those windows can overlap or fill time sparsely.